servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -6402,6 +6403,147 @@ void enforceGroupUserAuthorityRestrictions(final String domainName, final String
+                groupMember = new GroupMember();
+                groupMember.setMemberName(updatedUser.getFullName());
+                if (suspended) {
+                    groupMember.setSystemDisabled(currentGroup.getSystemDisabled() & ~Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                } else {
+                    groupMember.setSystemDisabled(currentGroup.getSystemDisabled() | Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                }
+                groupMembersWithUpdatedState = new ArrayList<>();
+                groupMembersWithUpdatedState.add(groupMember);
+
+                // Following method does Audit entry as well
+                if (updateGroupMemberDisabledState(null, con, groupMembersWithUpdatedState, currentGroup.getDomainName(),
+                        currentGroup.getGroupName(), ZMSConsts.SYS_AUTH_MONITOR, AUDIT_REF, caller)) {
+                    updatedGroups.add(currentGroup);
+                }
+            }
+            for (GroupMember updatedGM : updatedGroups) {
+
+                // update our group and domain time-stamps, and invalidate local cache entry
+                con.updateGroupModTimestamp(updatedGM.getDomainName(), updatedGM.getGroupName());
+                con.updateDomainModTimestamp(updatedGM.getDomainName());
+                cacheStore.invalidate(updatedGM.getDomainName());
+            }
+        }
+    }
+
+    private void updateRoleMembershipsByPrincipalState(boolean suspended, String caller, ObjectStoreConnection con, Principal updatedUser) {
+        RoleMember roleMember;
+        List<RoleMember> roleMembersWithUpdatedState;
+        DomainRoleMember domainRoleMember;
+        Set<MemberRole> updatedRoles = new HashSet<>();
+
+        domainRoleMember = con.getPrincipalRoles(updatedUser.getFullName(), null);
+        if (!domainRoleMember.getMemberRoles().isEmpty()) {
+            for (MemberRole memberRole : domainRoleMember.getMemberRoles()) {
+                roleMember = new RoleMember();
+                roleMember.setMemberName(updatedUser.getFullName());
+                if (suspended) {
+                    roleMember.setSystemDisabled(memberRole.getSystemDisabled() & ~Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                } else {
+                    roleMember.setSystemDisabled(memberRole.getSystemDisabled() | Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                }
+                roleMembersWithUpdatedState = new ArrayList<>();
+                roleMembersWithUpdatedState.add(roleMember);
+
+                // Following method does Audit entry as well
+                if (updateRoleMemberDisabledState(null, con, roleMembersWithUpdatedState, memberRole.getDomainName(),
+                        memberRole.getRoleName(), ZMSConsts.SYS_AUTH_MONITOR, AUDIT_REF, caller)) {
+                    updatedRoles.add(memberRole);
```

Comment:
the use of the this does not look correct since we're still processing the domain multiple times. if the principal is part of 10 roles in a domain, then we're updating the role mod timestamp for 10 roles, but then we're doing the same thing for the same domain 10 times.

instead of painting a updatedRoles struct, we should maintain:

Set<String> updatedDomains;

and then in the function we just call

con.updateRoleModTimestamp(memberRole.getDomainName(), memberRole.getRoleName());
updatedDomains.put(memberRole.getDomainName()); // this will ignore all duplicates

then outside of the for loop you can just iterate through the domain names and call

con.updateDomainModTimestamp(updatedMR.getDomainName());
cacheStore.invalidate(updatedMR.getDomainName());